sched/db_dump.cpp

```diff
@@ -209,32 +210,172 @@ int DUMP_SPEC::parse(FILE* in) {
     return ERR_XML_PARSE;
 }
 
+//
+// File streams
+//
+class OutputStream {
+public:
+    virtual ~OutputStream() {}
+    virtual bool is_open() const = 0;
+    virtual bool open(const char* filename) = 0;
+    virtual bool close() = 0;
+    virtual void write(const char* fmt, va_list args) = 0;
+};
+
+class UncompressedFile : public OutputStream
+{
+private:
+    FILE* f;
+
+public:
+    UncompressedFile()
+        : f(0) {}
+
+    bool is_open() const{
+        return f != 0;
+    }
+
+    bool open(const char* filename) {
+        f = fopen(filename, "w");
+        return f != 0;
+    }
+
+    bool close() {
+        if(!is_open())
+            return false;
+
+        fclose(f);
+        f = 0;
+        return true;
+    }
+
+    void write(const char* fmt, va_list args) {
+        if(!is_open())
+            return;
+
+        vfprintf(f, fmt, args);
+    }
+};
+
+class ZipFile : public OutputStream
+{
+private:
+    UncompressedFile f;
+    char current_path[MAXPATHLEN];
+
+public:
+    bool is_open() const {
+        return f.is_open();
+    }
+
+    bool open(const char* filename) {
+        if(!f.open(filename))
+            return false;
+
+        safe_strcpy(current_path, filename);
+        return true;
+    }
+
+    bool close() {
+        if(!f.close())
+            return false;
+
+        // Do zip
+        char buf[256];
+        sprintf(buf, "zip -q %s", current_path);
+        int retval = system(buf);
+        if (retval) {
+            log_messages.printf(MSG_CRITICAL,
+                "%s failed: %s\n", buf, boincerror(retval)
+            );
+            exit(retval);
+        }
+
+        return true;
+    }
+
+    void write(const char* fmt, va_list args) {
+        f.write(fmt, args);
+    }
+};
+
+class GzipFile : public OutputStream
+{
+private:
+    gzFile gz;
+
+public:
+    GzipFile()
+        : gz(0) {}
+
+    bool is_open() const {
+        return gz != 0;
+    }
+
+    bool open(const char* filename) {
+        char buf[256];
+        sprintf(buf, "%s.gz", filename);
+        gz = gzopen(buf, "wb");
+        return gz != 0;
+    }
+
+    bool close() {
+        if(!is_open())
+            return false;
+
+        gzclose(gz);
+        gz = 0;
+        return true;
+    }
+
+    void write(const char* fmt, va_list args) {
+        gzvprintf(gz, fmt, args);
```

Comment:
`gzvprintf()` is relatively new addition. RHEL 6 users are not going to like you using it and I don't think RHEL 7 users either. Another problem is that it uses 8 kB buffer and at least team descriptions can be larger than that.

Reply:
Good point. I considered the small buffer but discarded it as an issue as I only saw small writes. I'll have another look at it to see what the alternatives are.

Reply:
How about using `vasprintf` and then dumping the entire buffer to libz? It's going to be a bit slower but you are not bound to a buffer size.

Out of curiosity I double checked printf and its min buffer size is apparently 4095 character.